hbase_storage/src/main/java/com/splicemachine/storage/ClientPartition.java

```diff
@@ -155,32 +152,26 @@ public boolean checkAndPut(byte[] key,byte[] family,byte[] qualifier,byte[] expe
             List<Partition> partitions;
             if (!refresh) {
                 partitions = partitionInfoCache.getIfPresent(tableName);
-                if (partitions == null) {
-                    partitions = formatPartitions(getAllRegionLocations(false));
-                    assert partitions!=null:"partitions are null";
-                    partitionInfoCache.put(tableName, partitions);
+                if (partitions != null) {
+                    return partitions;
                 }
-                return partitions;
             }
-            partitions = formatPartitions(getAllRegionLocations(true));
-            partitionInfoCache.invalidate(tableName);
-            assert partitions!=null:"partitions are null";
-            partitionInfoCache.put(tableName,partitions);
+            List<HRegionLocation> tableLocations = ((ClusterConnection) connection).locateRegions(tableName, !refresh, false);
+            partitions = new ArrayList<>(tableLocations.size());
+            for (HRegionLocation location : tableLocations){
+                if (location.getServerName() == null) {     // ensure no offline regions
+                    continue;
```

Comment:
Is it OK to ignore offline regions?  The old code kept retrying until all regions were online.  Could we potentially end up not scanning all rows in the table?

Reply:
_HRegionLocation.getServerName() == null_ causes the NPE. I believe it's an issue in HBase itself as serverName is used in HRegionLocation.hashCode() and throws NPE when null (making such HRegionLocation instances poorly defined and unusable).
In fact, _ClusterConnection.locateRegions(..., false)_ already ensures that such locations are filtered out. I make that condition explicit..

Reply:
I see, thanks.